Fix nullptr deref in B3::ReduceStrength's handling of Trunc for double constants.

MenloDorian opened the pull request:
#### ab0808f76fdb011db0cb037137d236849534ac26
<pre>
Fix nullptr deref in B3::ReduceStrength&apos;s handling of Trunc for double constants.
<a href="https://bugs.webkit.org/show_bug.cgi?id=265730">https://bugs.webkit.org/show_bug.cgi?id=265730</a>
<a href="https://rdar.apple.com/116459635">rdar://116459635</a>

Reviewed by Yusuke Suzuki.

We cannot run asInt64() on a ConstDoubleValue.  This patch fixes this by handling
ConstDoubleValue separately and running asDouble() on it instead.

* Source/JavaScriptCore/b3/B3ReduceStrength.cpp:

Canonical link: <a href="https://commits.webkit.org/271580@main">https://commits.webkit.org/271580@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/25b32e467c596dcb020fe34e8494ef297ca9657f

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/28492 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/7137 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/29878 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/31019 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/25947 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/9238 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/4506 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/26208 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/28761 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/5893 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/24519 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/5156 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/5274 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/25518 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/31705 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/24655 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/26093 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/25961 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/31552 "Passed tests") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/27586 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/5234 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/3410 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/29329 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/6835 "Built successfully") | | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/35110 "Built successfully") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/5691 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/7587 "Passed tests") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/3721 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/5754 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

Constellation commented on Source/JavaScriptCore/b3/B3ReduceStrength.cpp:
```diff
@@ -1982,8 +1982,10 @@ class ReduceStrength {
             if (m_value->child(0)->hasInt64() || m_value->child(0)->hasDouble()) {
-                replaceWithNewValue(
-                    m_proc.addIntConstant(m_value, static_cast<int32_t>(m_value->child(0)->asInt64())));
+                int32_t truncValue = m_value->child(0)->hasInt64()
```
This looks not correct from how `Trunc(Double)` works from b3/intermediate-representation.html. See 2ea19e682e31953ec3116b53140f146a7458106c

Constellation commented on Source/JavaScriptCore/b3/B3ReduceStrength.cpp:
```diff
@@ -1981,7 +1981,12 @@ class ReduceStrength {
+            // The only time this is meaningful to do to a double value is when the value is in an FP
+            // argument register, which by definition is not constant. Hence, it does not make sense
+            // for look for a Trunc on a ConstantDoubleValue. The only type of constant of relevance
+            // here is Int64.
```
I think we should have a change like,

```
if (m_value->child(0)->hasDouble()) {
    double value = m_value->child(0)->asDouble();
    replaceWithNewValue(m_proc.addConstant(m_value->origin(), m_value->type(), bitwise_cast<int64_t>(value)));
    break;
}
```

Constellation reviewed (approved):
r=me

Commit:
Fix nullptr deref in B3::ReduceStrength's handling of Trunc for double constants.
https://bugs.webkit.org/show_bug.cgi?id=265730
rdar://116459635

Reviewed by Yusuke Suzuki.

We cannot run asInt64() on a ConstDoubleValue.  This patch fixes this by handling
ConstDoubleValue separately and running asDouble() on it instead.

* Source/JavaScriptCore/b3/B3ReduceStrength.cpp:

Canonical link: https://commits.webkit.org/271580@main
